feat: Pass remote Markdown images through image service

p0lyw0lf opened the pull request:
## Changes

+ Currently, in Markdown files, only local images get passed to the image service
  + This was because the image service only supported local files when it was first released.
+ However, the image service now supports remote images
+ So, this PR adds support for passing remote images in Markdown files to the image service.

## Testing

I tested this change by creating a small project with a single allowed remote domain, and rendering markdown files from:

+ a page
+ a content collection

Both approaches work. I can provide these test files if needed, though I'm not sure how we'll want to deal with the fact it'll require an internet connection.

### Known Issues

If the remote image is styled in some way, the newly added `width` and `height` attributes could mess up that styling. Ideally, we'd like to be able to modify the properties after `getImage` runs somehow, but seeing as this is also the case for local images transformed using this method too, I think this is more a deficiency in `getImage` best left addressed by the responsive image experiment.

This is only an issue for remote images as configured under `images.domains`. Other images are untouched.

## Docs

If this change goes through, we will need to update the images guide to reflect that remote images are now supported in markdown files. PR: https://github.com/withastro/docs/pull/11021

ematipico commented on packages/astro/src/content/runtime.ts:
```diff
@@ -414,13 +414,20 @@ async function updateImageReferencesInBody(html: string, fileName: string) {
-			if (!id || imageObjects.has(id) || !imported) {
-				continue;
+			let image: GetImageResult;
+			if (validURL(decodedImagePath.src)) {
```
You can use [`URL.canParse`](https://developer.mozilla.org/en-US/docs/Web/API/URL/canParse_static)

ematipico commented on packages/astro/src/content/runtime.ts:
```diff
@@ -414,13 +414,20 @@ async function updateImageReferencesInBody(html: string, fileName: string) {
-				continue;
+			let image: GetImageResult;
+			if (validURL(decodedImagePath.src)) {
+				// Remote image, pass through without resolving import
+				image = await getImage(decodedImagePath);
```
Shouldn't we validate if the remote image we are about to download is part of the allowed domains? https://docs.astro.build/en/guides/images/#authorizing-remote-images

ematipico commented on packages/markdown/remark/src/remark-collect-images.ts:
```diff
@@ -1,42 +1,45 @@
+
+			try {
+				const domain = new URL(url).host;
+				if (allowedRemoteDomains.has(domain)) remoteImagePaths.add(url);
+			} catch {
+				// Not a valid remote URL. Check if it's a local image. If it's an absolute path, then it's not.
+				if (!url.startsWith('/')) localImagePaths.add(url);
```
A couple of comments:
- nit: can you stick to our internal style and use brackets after if statements?
- use `URL.canParse` instead of relying on a try/catch

ematipico commented on packages/markdown/remark/src/rehype-images.ts:
```diff
@@ -22,6 +22,21 @@ export function rehypeImages() {
 
 						node.properties['__ASTRO_IMAGE_'] = JSON.stringify({ ...props, index });
 
+						Object.keys(props).forEach((prop) => {
+							delete node.properties[prop];
+						});
+					} else if (file.data.astro?.remoteImagePaths?.includes(node.properties.src)) {
+						const { ...props } = node.properties;
+
+						const index = imageOccurrenceMap.get(node.properties.src) || 0;
+						imageOccurrenceMap.set(node.properties.src, index + 1);
+
+						node.properties['__ASTRO_IMAGE_'] = JSON.stringify({
+							inferSize: 'width' in props && 'height' in props ? undefined : true,
+							...props,
+							index,
+						});
+
```
This new `else if` branch has code that is repeated to the existing `if` branch, can we find a better way to not have repeated code?

ematipico commented on packages/integrations/mdx/src/rehype-images-to-component.ts:
```diff
@@ -73,84 +73,105 @@ function getImageComponentAttributes(props: Properties): MdxJsxAttribute[] {
 		visit(tree, 'element', (node, index, parent) => {
-			if (!file.data.astro?.imagePaths?.length || node.tagName !== 'img' || !node.properties.src)
-				return;
+			if (node.tagName !== 'img' || !node.properties.src) return;
```
Why was this changed with fewer checks? Shouldn't we add a check for remote images?

ematipico reviewed (commented):
Thank you @p0lyw0lf for your PR! I left some comments and questions around the new feature

p0lyw0lf commented:
@ematipico I believe I've addressed your feedback. One of the bigger changes was moving the checks for https://docs.astro.build/en/guides/images/#authorizing-remote-images out of `astro` and into `@astrojs/markdown-remark`, so as to not duplicate them.

p0lyw0lf commented:
The fixes I pushed this morning were not enough; there was a error where moving `isRemoteAllowed` into `@astrojs/markdown-remark` and including in `astro` caused `node:fs` to show up in the module path somehow. I fixed this by moving that code into `@astrojs/internal-helpers` instead. There was also a small error with that move in the first place, which I have also fixed. Tests pass locally now, hoping they pass here too.

ematipico commented:
Since you added code to another package, that package must have a changeset. Without changeset it won't have its version bumped and it will break in production.

It should be a new changeset, not the one you currently have

p0lyw0lf commented:
@ematipico I've updated the changeset

ematipico commented:
@p0lyw0lf Thank you for you PR. Coding-wise, it looks good to me. Would you be able to send a PR to docs? If not, that's fine, we can do that. Just let us know. We aim to release the new feature next Thursday.

p0lyw0lf commented on packages/astro/src/vite-plugin-markdown/images.ts:
```diff
@@ -29,6 +33,25 @@ export function getMarkdownCodeForImages(imagePaths: MarkdownImagePath[], html:
+											)} + '[^"]*)"', 'g');
+											let match;
+											let occurrenceCounter = 0;
+											while ((match = regex.exec(html)) !== null) {
```
The `localImagePaths` are processed the exact same way. I would prefer to fix how this is done in a follow-up PR if that is alright with you?

p0lyw0lf commented:
@ematipico I opened a docs PR: https://github.com/withastro/docs/pull/11021

sarah11918 commented on .changeset/shy-bats-exist.md:
```diff
@@ -0,0 +1,17 @@
+'astro': minor
+---
+
+Pass remote Markdown images through image service
+
+Previously, Astro only allowed local images to be optimized when included using
+`![]()` syntax in plain Markdown files. This was because, when the image
+service was first introduced, it only was able to optimize those images. Now,
+however, Astro's image service can optimize remote images as well. So, we can
+add support for this!
+
+This is a semver-minor bump because it can significantly change what's output
+under certain circumstances.
```
I *think* this needs to be a different changeset for each package changed because different things were changed in each package. (e.g. the MDX changset can look similar to the main Astro one, but the changeset for the other packages should probably describe exactly what about each of them was changed) 

Ema can guide you on that, and I can look at the wording after they're done!

But I'll suggest some wording for the `astro` main changeset, something kind of like:

```suggestion
Adds the ability to process and optimize remote images in Markdown files

Previously, Astro only allowed local images to be optimized when included using `![]()` syntax in plain Markdown files. Astro's image service could only display remote images without any processing. 

Now, Astro's image service can also optimize remote images written in standard Markdown syntax. This allows you to enjoy the benefits of Astro's image processing when your images are stored externally, for example in a CMS or digital asset manager.

No additional configuration is required to use this feature! Any existing remote images written in Markdown will now automatically be optimized. To opt-out of this processing, write your images in Markdown using the HTML `<img>` tag instead. Note that images located in your `public/` folder are still never processed.
```

(Something similar to above but referring specifically to "Markdown syntax in MDX files" instead of "Markdown files" would be my suggestion for that changeset.)

sarah11918 reviewed (commented):
Exciting feature! 🚀 

I left my thought on the changeset for docs, and Ema can guide you on creating individual ones that describe what's new / changes about the other packages.

I've also reviewed the docs PR which won't take much to polish it up! 🙌

p0lyw0lf commented:
@sarah11918 @ematipico I've split the changeset into multiple, adding specifics on what's changed for each package. Thanks again for reviewing this!!

sarah11918 commented on .changeset/tiny-cows-march.md:
```diff
@@ -0,0 +1,11 @@
+
+Now, Astro's image service can also optimize remote images written in standard Markdown syntax. This allows you to enjoy the benefits of Astro's image processing when your images are stored externally, for example in a CMS or digital asset manager.
+
+No additional configuration is required to use this feature! Any existing remote images written in Markdown will now automatically be optimized. To opt-out of this processing, write your images in Markdown using the HTML `<img>` tag instead. Note that images located in your `public/` folder are still never processed.
```
```suggestion
No additional configuration is required to use this feature! Any existing remote images written in Markdown will now automatically be optimized. To opt-out of this processing, write your images in Markdown using the JSX `<img/>` tag instead. Note that images located in your `public/` folder are still never processed.
```
Just checking, but I think MDX requires the JSX self-closing image tag?

sarah11918 commented on .changeset/warm-planes-swim.md:
```diff
@@ -0,0 +1,14 @@
+'@astrojs/markdown-remark': minor
+---
+
+Transform remote images in addition to local images
+
+Previously, an internal remark plugin only looked for images in `![]()` syntax
+that referred to a relative path, passing through to an internal rehype plugin
+that would transform them for later processing by Astro's image service.
+
+Now, the plugins transform both local and remote images, outputting them into
+`localImagePaths` and `remoteImagePaths` metadata fields. A new configuration
+option, mirroring the one used by Astro, can be provided to control which
+remote images get processed this way.
```
Great, thanks for the clarification! 

We do already say this in docs:

> For extra protection when dealing with external sources, Astro’s image components and helper function will only process (e.g. optimize, transform) images from [authorized image sources specified in your configuration](https://docs.astro.build/en/guides/images/#authorizing-remote-images). Remote images from other sources will be displayed with no processing.

And then have a whole section on authorizing remote images.

Is what you're saying above anything different from what we already said about remote images needing to be authorized for any optimization? If so, I don't think a *user* needs to know anything more than that? If it *always* is needed, then it's also needed in this case? Is there anything else in particular to add?

p0lyw0lf replied:
It's true, users of `astro` wouldn't need to know anything about this, because that configuration is passed from their Astro config automatically.

I was thinking that users of `@astrojs/markdown-remark` as a standalone library might like to know about this, though maybe since it's intended to be internal-only it doesn't make sense to mention.

sarah11918 replied:
Yeah, I *think* we don't really expect/optimize for any users of this as a standalone library? Perhaps @ematipico or @ascorbic will weigh in here.

But, it could totally be valid to consider people who might *contribute* to this package and note something like "works based on the project's configured authorized domains" or something like that?

p0lyw0lf replied:
I've added your changes above, plus further wording based on what's in the docs.

sarah11918 reviewed (approved):
Looks good to me! Approving for docs! 🥳 

(And noting that I dont see any new error messages introduced with this feature)

aholbreich commented:
Has anything changed on the usage side?

having
---
cover: https://remote-cdn.io/cool.jpeg
---
would still work? 
Btw. I remember to see some warnings about `width `in those cases..
